tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+
+    // Threads that will not be doing anything useful when reading the matrix
+    // because the thread block size is bigger than the data block size.
+    int wasted_threads = (data_long_side -
+                          MathUtil::FloorOfRatio<int>(
+                              data_long_side, proposed_tile_long_side_len) *
+                              proposed_tile_long_side_len);
+
+    int num_full_tiles = MathUtil::FloorOfRatio<int>(
+        data_long_side, proposed_tile_long_side_len);
+
+    float cost = 0;
+
+    // However, if we can execute two or more full tiles, then we gladly
+    // accept any number of wasted thread and ignore its cost.
+    if (num_full_tiles <= 1) cost = wasted_threads;
+
+    // Using less and equal here because given the same cost, we would like to
+    // launch as many threads as possible.
+    if (cost <= lowest_cost) {
+      tile_long_side_len = proposed_tile_long_side_len;
+      tile_short_side_len = tile_size_pair.second;
+      lowest_cost = cost;
+    }
+  }
+
+  // Request tile sizes such that the longer side of threadblock aligns with
+  // the longer side of input data block to maximize read throughput.
+  // The ideal tile shape to request is one with its length of the shorter
+  // side of the tile being equal to the length of the shorter side of the
+  // input matrix.
+  int requested_tile_size_i = input_dims[1] >= kMinDimensionToUseTiles
+                                  ? tile_long_side_len
+                                  : input_dims[1];
+  int requested_tile_size_j = input_dims[1] >= kMinDimensionToUseTiles
+                                  ? input_dims[2]
+                                  : tile_long_side_len;
+
+  // Truncate the shorter size requested according to the manual limit set in
+  // tile_spec to make sure that we do not launch configurations violating
+  // hardware limits.
+  requested_tile_size_i = requested_tile_size_i == tile_long_side_len
+                              ? tile_long_side_len
+                              : min(requested_tile_size_i, tile_short_side_len);
+  requested_tile_size_j = requested_tile_size_j == tile_long_side_len
+                              ? tile_long_side_len
+                              : min(requested_tile_size_j, tile_short_side_len);
+
+  Dimension<3> input_dims_in_tiles = {
+      input_dims[0],
+      (input_dims[1] + requested_tile_size_i - 1) / requested_tile_size_i,
+      (input_dims[2] + requested_tile_size_j - 1) / requested_tile_size_j,
+  };
+
+  int total_tiles_count =
+      input_dims_in_tiles[0] * input_dims_in_tiles[1] * input_dims_in_tiles[2];
+
+#define DISPATCH_WITH_UINT_TYPE(TYPE_SIZE)                                \
+  BatchNarrowMatrixTransposeDispatcher<uint##TYPE_SIZE, 32, 2>::DoIt(     \
+      d, requested_tile_size_i, requested_tile_size_j, total_tiles_count, \
+      reinterpret_cast<const uint##TYPE_SIZE*>(input), input_dims,        \
+      reinterpret_cast<uint##TYPE_SIZE*>(output))
```

Comment:
Perhaps it would be worthwhile to have something like:

```
// TransposeElemType<N>::type is an N-byte type.  We use this to map types of the same size onto a single type -- for example, {int32, uint32, and float} all map to ElemType<4>::type.  This lets us compile just one transpose kernel for all N-byte types.
template <int ElemBytes> struct TransposeElemType;
template <> struct TransposeElemType<1> { using type = uint8; };
template <> struct TransposeElemType<2> { using type = uint16; };
template <> struct TransposeElemType<4> { using type = uint32; };
template <> struct TransposeElemType<8> { using type = uint64; };
template <> struct TransposeElemType<16> { using type = float4; };
```

Then we don't need this macro or switch statement: it becomes simply:

```
using ElemType = TransposeElemType<sizeof(T)>::type;
static_assert(alignof(T) >= alignof(ElemType));
BatchNarrowMatrixTransposeDispatcher<ElemType, 32, 2>::DoIt(...);
```

WDYT?